### PR TITLE
8272317: jstatd has dependency on Security Manager which needs to be removed

### PR DESCRIPTION
**Pull request by kevinjwalls**
Remove the use of Security Manager from jstatd.
Add use of an ObjectInputFilter to restrict RMI.

Also we can undo the property-setting Launcher.gmk change from: 8279007: jstatd fails to start because SecurityManager is disabled
..as that is no longer needed.

Docs/man page update to follow (JDK-8278619).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8272317](https://bugs.openjdk.java.net/browse/JDK-8272317): jstatd has dependency on Security Manager which needs to be removed
 * [JDK-8279891](https://bugs.openjdk.java.net/browse/JDK-8279891): jstatd has dependency on Security Manager which needs to be removed (**CSR**)

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to e49b9fe24caf860d99bcf15046e5cb287ab9ddf9
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to e49b9fe24caf860d99bcf15046e5cb287ab9ddf9

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6919/head:pull/6919` \
`$ git checkout pull/6919`

Update a local copy of the PR: \
`$ git checkout pull/6919` \
`$ git pull https://git.openjdk.java.net/jdk pull/6919/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6919`

View PR using the GUI difftool: \
`$ git pr show -t 6919`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6919.diff">https://git.openjdk.java.net/jdk/pull/6919.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jstatd/share/classes/sun/tools/jstatd/Jstatd.java [hunk 1]
```diff
@@ -47,6 +48,8 @@
     private static boolean startRegistry = true;
     private static RemoteHost remoteHost;
 
+    private static final String rmiFilterPattern = "sun.jvmstat.monitor.remote.RemoteVm;com.sun.proxy.jdk.proxy1.$Proxy1;com.sun.proxy.jdk.proxy1.$Proxy2;java.lang.reflect.Proxy;java.rmi.server.RemoteObjectInvocationHandler;java.rmi.server.RemoteObject;!*";
```
Comment (mlchung) [inline src/jdk.jstatd/share/classes/sun/tools/jstatd/Jstatd.java:51]: The class name of the dynamic proxy is generated at runtime and can be different.   As Bernd commented, the proxy classes cannot/should not be listed in the filter pattern.

Comment (kevinjwalls) [reply src/jdk.jstatd/share/classes/sun/tools/jstatd/Jstatd.java:51]: OK thanks - I was trying the minimal pattern to overcome rejections such as the following, captured in logs on different runs:

<message>ObjectInputFilter REJECTED: class com.sun.proxy.jdk.proxy1.$Proxy1, array length: -1, nRefs: 2, depth: 1, bytes: 84, ex: n/a</message>

<message>ObjectInputFilter REJECTED: class com.sun.proxy.jdk.proxy1.$Proxy2, array length: -1, nRefs: 2, depth: 1, bytes: 84, ex: n/a</message>

Comment (stuart-marks) [reply src/jdk.jstatd/share/classes/sun/tools/jstatd/Jstatd.java:51]: I think the proxy classes need to be there. The `RemoteHost` API has a parameter of type `RemoteVm` which is a stub to an RMI remote object, which consists of a proxy and a handler. The proxy's interface list is filtered by the serialization filter so somebody can't just pass a proxy for anything.

The name of the proxy class probably does need to be wildcarded though.

**Comment by kevinjwalls** [comment]
Thanks for the comments - 

The proxy objects are needed in the filter for this to work at all.

The proxy names/numbers and innner class names/numbers are predictable and stable IF we are using jstatd and jstat as standalone programs.

But they are unstable if there is other relevant activity in the VM process, e.g. a JMX connection comes in before they are created.  We should wildcard the proxy names to work in such a situation:  com.sun.proxy.jdk.proxy*

I'll mention also that jstatd has always been an experimental feature.  The man page warns about lack of authentication and advises usage with caution.

File: src/jdk.jstatd/share/classes/sun/tools/jstatd/Jstatd.java (hunk 1 above)
Comment (RogerRiggs) [reply src/jdk.jstatd/share/classes/sun/tools/jstatd/Jstatd.java:51]: @mlchung The Proxy class passed to the filter has been created in this VM from the interfaces listed.
The interfaces have already been filtered prior to creating the proxy.
The Proxy classes can safely be allowed based on a wildcard of the name. (As Stuart said).

**Comment by kevinjwalls** [comment]
CSR has been approved (https://bugs.openjdk.java.net/browse/JDK-8279891)

**Comment by mlchung** [comment]
Are all the proxy interfaces public?    The package in which a proxy class is created may be different depending if whether any proxy interface is in a non-exported and non-open package.   `com.sun.proxy.jdk.proxy*` is the package for non-exported proxy classes.  The proxy classes may be public in an unconditionally exported package [1] and its package name is `jdk.proxy*`.

[1] https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/lang/reflect/Proxy.html#membership

**Comment by kevinjwalls** [comment]
> Are all the proxy interfaces public?

sun.jvmstat.monitor.remote.RemoteVm is "public interface RemoteVm extends Remote" and methods in there only return basic types.
This is in the jdk.jstatd module, where I see the module info contains "exports sun.jvmstat.monitor.remote to java.rmi;"

The only other names permitted are proxy/reflect/rmi related.

**Comment by kevinjwalls** [comment]
> If `sun.jvmstat.monitor.remote.RemoteVm` is the only proxy interface, `com.sun.proxy.jdk.proxy*` should adequately cover the proxy classes created for `RemoteVm`.

Thanks.  With that endorsement I think there are no unresolved issues with this change.

**Review by plummercj — APPROVED** [review]

File: test/jdk/sun/tools/jstatd/JstatdTest.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2013, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (RogerRiggs) [inline test/jdk/sun/tools/jstatd/JstatdTest.java:2]: Update copyrights.

**Review by RogerRiggs — APPROVED** [review]
LGTM

File: test/jdk/sun/tools/jstatd/JstatdTest.java (hunk 2 above)
Comment (kevinjwalls) [reply test/jdk/sun/tools/jstatd/JstatdTest.java]: Yes, thanks, time has moved on since I started!...

**Comment by kevinjwalls** [comment]
Many thanks Chris, Roger!

**Pull request closed without merging**